### ISSUE TITLE
MAUI Dependencies cause issues when updating the MAUI workload with 3rd party dependencies

### ISSUE THREAD

**Issue by dansiegel** [dotnet/maui#8032]
### Description

The issue was first reported by @rdavisau in the SponsorConnect Discord. After updating to MAUI SR1 a project with 3rd party dependencies built against MAUI GA caused a NuGet Dependency downgrade that broke the build. This seems to be caused by a NuGet dependency which is added to libraries which have `UseMaui`. As discussed with @Redth these dependencies probably shouldn't be added to NuGet packages built against MAUI

![image](https://user-images.githubusercontent.com/3860573/173722659-acfda301-6f2c-4665-978c-6947f169322f.png)
> NuGet Dependencies that are automatically added

![image](https://user-images.githubusercontent.com/3860573/173722877-1f61e9f1-cbb6-41f7-8a69-e1cb92e3362e.png)
> Resulting error after updating to SR1

Reproduction: https://github.com/rdavisau/maybe-update-maui

### Steps to Reproduce

1. Create a MAUI app
2. Add Community Packages that were built against the MAUI GA
3. Use SR1

### Version with bug

6.0 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

iOS, Android, Windows, macOS, Other (Tizen, Linux, etc. not supported by Microsoft directly)

### Affected platform versions

all

**Comment by egvijayanand** [comment]
Yes, faced the same issue in [AlohaKit](https://www.nuget.org/packages/AlohaKit) NuGet package that was released well before the release of SR1.

Couldn't reference the package in the GA version of the .NET MAUI App. Thought the author wanted to take advantage of the pre-release features but seems to be this issue.

**Comment by Redth** [comment]
For now here's a workaround you can use in your library projects to rebuild them without containing the dependency.  Add this to your library project which you are packaging into a nuget:

```
<ItemGroup>
	<PackageReference Include="Microsoft.Maui.Dependencies" Version="$(MauiVersion)">
		<PrivateAssets>All</PrivateAssets>
	</PackageReference>
	<PackageReference Include="Microsoft.Maui.Extensions" Version="$(MauiVersion)">
		<PrivateAssets>All</PrivateAssets>
	</PackageReference>
</ItemGroup>
```

**Comment by egvijayanand** [comment]
Have included the aforementioned `ItemGroup` definition then updated the MAUI version to the GA release and then commented out the `<UseMaui>true</UseMaui>` tag in the project file.

But now the issue is namespaces for controls like ContentPage, StackLayout, ... is not getting resolved.

When `Microsoft.Maui.Controls` namespace is added manually results in the below error.

The type or namespace name `Controls` does not exist in the namespace `Microsoft.Maui` (are you missing an assembly reference?).

Am I missing anything here?

**Comment by egvijayanand** [comment]
And the point to note here is that `Microsoft.Maui` and `Microsoft.Maui.Graphics` namespace and associated types are getting resolved without any issues.

**Comment by Redth** [comment]
Why are you commenting out UseMaui? That still needs to be there.

**Comment by egvijayanand** [comment]
With that option enabled, the packages restored with the latest version.

Anyway let me try again and update.

**Comment by egvijayanand** [comment]
Have tried both the UseMaui option and the explicit NuGet package references, now the project is building without errors but while analyzing the generated NuGet output, there's no reference to the .NET MAUI packages.

<img width="275" alt="dotnetmaui-with-explicit-package-reference" src="https://user-images.githubusercontent.com/81947404/174082996-f0836526-451a-4f2c-b71d-e5bdc49dc382.png">

Whereas the conventional project (without any explicit package references), the NuGet output has its references set to the latest SR1 version.

<img width="281" alt="dotnetmaui-without-package-reference" src="https://user-images.githubusercontent.com/81947404/174083239-e44025ac-5fcb-415b-b82b-c0efa987a3f4.png">

Is this the expected behavior? How to generate libraries that target the baselined version, GA release in this case?

**Comment by egvijayanand** [comment]
At last, found the solution for this issue in an [article](https://dev.to/davidortinau/installing-net-maui-on-macos-4mmc) by @davidortinau

Define the below-mentioned tag in the project file along with the version of the .NET MAUI to be used during package restoration.

```xml
<PropertyGroup>
    <UseMaui>true</UseMaui>
    <!-- Define the version of .NET MAUI to be used for package restore -->
    <MauiVersion>6.0.312</MauiVersion>
</PropertyGroup>
```
And the build output also reflects the right dependency version. 

Can these things be documented somewhere in the repository?

<img width="274" alt="dotnetmaui-with-mauiversion-tag" src="https://user-images.githubusercontent.com/81947404/174456000-bab1eabd-e2d2-4c8e-9129-0d97b2a4317d.png">

**Comment by mattleibow** [comment]
@egvijayanand did you try the workaround by @Redth ?

```xml
<ItemGroup>
    <PackageReference Include="Microsoft.Maui.Dependencies" Version="$(MauiVersion)" PrivateAssets="all" />
    <PackageReference Include="Microsoft.Maui.Extensions" Version="$(MauiVersion)" PrivateAssets="all" />
</ItemGroup>
```

It worked for me... No need to change versions, just tell dotnet not to add a reference to those packages...

**Comment by egvijayanand** [comment]
Refer to my penultimate comment, that didn't work as expected.

But setting the `MauiVersion` alone in the project file made it work as expected.

**Comment by mattleibow** [comment]
But that is correct. There _should_ be no reference to Maui from NuGet packages. It is all implicit.

**Comment by egvijayanand** [comment]
I'm not sure about the happenings currently underway.

But till now, the library packages that are getting published to NuGet have references to the MAUI packages as their dependencies.

Setting the `MauiVersion` tag pointing to the GA version makes the minimum version required to that baselined version even when built on top of **SR1** so that users on the earlier version can also make use of the packages.

But once any new API introduced as part of the recent version is made use of, then the dependent version can be moved forward.

**Comment by mattleibow** [comment]
> But till now, the library packages that are getting published to NuGet have references to the MAUI packages as their dependencies.

This is incorrect behavior. The PR will fix that.

> Setting the MauiVersion tag pointing to the GA version

This is not meant to be used generally and is only for testing nightly builds right now.

**Comment by egvijayanand** [comment]
Providing a few examples.

https://www.nuget.org/packages/Prism.Maui/8.1.228-beta#dependencies-body-tab

The below version of the package is built on top of SR1 but the works even on top of the baselined GA version.

https://www.nuget.org/packages/VijayAnand.MauiToolkit#dependencies-body-tab

**Comment by mattleibow** [comment]
I understand what you observe. But you are observing incorrect and undesirable behaviour.

**Comment by egvijayanand** [comment]
> I understand what you observe. But you are observing incorrect and undesirable behaviour.

If these dependencies are not mentioned, then how does package resolution work moving forward?

**Comment by egvijayanand** [comment]
Also, the library maintainer should have little control over on top of which version their libraries are getting released to reach a broader set of audiences and more importantly stability.

**Comment by mattleibow** [comment]
> If these dependencies are not mentioned, then how does package resolution work moving forward?

The `<UseMaui>true</UseMaui>` controls whether or not to reference the maui framework.

> the library maintainer should have little control over on top of which version their libraries are getting released

True, but since the API of Maui is stable and will not change, it does not matter which version you build against because it will always be compatible with all net6.0 versions of maui.

Also, the MauiVersion property is deceiving because it does not actually do anything that you expect and just controls the version of the dependencies library. It used to do more, but since moving to workloads it has a far reduced power.

**Comment by egvijayanand** [comment]
> True, but since the API of Maui is stable and will not change, it does not matter which version you build against because it will always be compatible with all net6.0 versions of maui.

That can't be the case, API sets do tend to change, either introduced newly or fixed to address issues.

Say SR1 broke the [PushModalAsync](#8159 ) API of the Navigation page, again am not complaining, but changes are imminently visible.

A library maintainer cannot do complete testing of all the APIs for every release of MAUI as it'll it involves time and effort unless there's a compelling reason to move to that release, say security fixes.

There should be an option to stick to a baseline version.

**Comment by mattleibow** [comment]
> There should be an option to stick to a baseline version.

Possibly, but MauiVersion is not the way today and the correct fix is to remove the references. Once https://github.com/dotnet/maui/pull/8064 is merged, the maui references will no longer appear and I am hoping this gets in for SR2.

**Comment by egvijayanand** [comment]
> > There should be an option to stick to a baseline version.
> 
> Possibly, but MauiVersion is not the way today and the correct fix is to remove the references. Once #8064 is merged, the maui references will no longer appear and I am hoping this gets in for SR2.

Will check it out once released. Request to add some documentation while releasing the same for reference.

**Comment by mattleibow** [comment]
I actually created this one to help get some clarity: https://github.com/dotnet/sdk/issues/26130

**Comment by mattleibow** [comment]
Closing this issues sine https://github.com/dotnet/maui/pull/8064 was merged.

**Comment by bijington** [comment]
> I actually created this one to help get some clarity: https://github.com/dotnet/sdk/issues/26130

@mattleibow Thank you. You captured far more information than I had supplied. Great work!

I will continue to monitor that new issue.

Thanks again
